fmriprep/workflows/fieldmap/base.py

```diff
@@ -36,48 +36,223 @@
+            fo the fieldmap-less SyN method
+        template : str
+            Name of template targeted by `'template'` output space
 
 
+    **Outputs**
+        bold_ref
+            An unwarped BOLD reference
+        bold_mask
+            The corresponding new mask after unwarping
+        bold_ref_brain
+            Brain-extracted, unwarped BOLD reference
+        out_warp
+            The deformation field to unwarp the susceptibility distortions
+        syn_bold_ref
+            If ``--force-syn``, an unwarped BOLD reference with this
+            method (for reporting purposes)
+
     """
 
-    # pybids type options: (phase1|phase2|phasediff|epi|fieldmap)
-    # https://github.com/INCF/pybids/blob/213c425d8ee820f4b7a7ae96e447a4193da2f359/bids/grabbids/bids_layout.py#L63
-    if fmap_bids['type'] == 'fieldmap':
-        from .fmap import init_fmap_wf
-        fmap_wf = init_fmap_wf(reportlets_dir=reportlets_dir,
-                               omp_nthreads=omp_nthreads,
-                               fmap_bspline=fmap_bspline)
-        # set inputs
-        fmap_wf.inputs.inputnode.fieldmap = fmap_bids['fieldmap']
-        fmap_wf.inputs.inputnode.magnitude = fmap_bids['magnitude']
-        return fmap_wf
-
-    if fmap_bids['type'] == 'phasediff':
-        from .phdiff import init_phdiff_wf
-        phdiff_wf = init_phdiff_wf(reportlets_dir=reportlets_dir,
-                                   omp_nthreads=omp_nthreads)
-        # set inputs
-        phdiff_wf.inputs.inputnode.phasediff = fmap_bids['phasediff']
-        phdiff_wf.inputs.inputnode.magnitude = [
-            fmap_ for key, fmap_ in sorted(fmap_bids.items())
-            if key.startswith("magnitude")
-        ]
-        return phdiff_wf
-
-    if fmap_bids['type'] in ['phase1', 'phase2']:
-        raise NotImplementedError
+    # TODO: To be removed (supported fieldmaps):
+    if not set([fmap['type'] for fmap in fmaps]).intersection(FMAP_PRIORITY):
+        fmaps = None
+
+    workflow = pe.Workflow(name='sdc_wf' if fmaps else 'sdc_bypass_wf')
+    inputnode = pe.Node(niu.IdentityInterface(
+        fields=['name_source', 'bold_ref', 'bold_ref_brain', 'bold_mask',
+                't1_brain', 't1_2_mni_reverse_transform', 'template']),
+        name='inputnode')
+
+    outputnode = pe.Node(niu.IdentityInterface(
+        fields=['bold_ref', 'bold_mask', 'bold_ref_brain',
+                'out_warp', 'syn_bold_ref']),
+        name='outputnode')
+
+    # No fieldmaps - forward inputs to outputs
+    if not fmaps:
+        workflow.connect([
+            (inputnode, outputnode, [('bold_ref', 'bold_ref'),
+                                     ('bold_mask', 'bold_mask'),
+                                     ('bold_ref_brain', 'bold_ref_brain')]),
+        ])
+        return workflow
+
+    # In case there are multiple fieldmaps prefer EPI
+    fmaps.sort(key=lambda fmap: FMAP_PRIORITY[fmap['type']])
+    fmap = fmaps[0]
```

Comment:
I feel like the selection logic is a bit distributed. I'd rather make decisions in one place, and then just work through the consequences. So here, I would look for two things: (1) Which primary SDC should we run? (2) Are we running SyN?

```Python
fmap = min(fmaps, key=lambda...)
run_syn = any(fm['type'] == 'syn' for fm in fmaps)
```

Yeah, it's iterating twice, but at `n < 4`, it doesn't matter. If you want to sort and use `fmaps[0]` and `fmaps[-1]`, that's fine, but I feel like this is slightly clearer in meaning.

Reply:
done at the only branching where the `run_syn` is tested, thanks!